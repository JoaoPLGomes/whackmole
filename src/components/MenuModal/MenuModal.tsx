import { INITIAL_TIMER } from '@/constants/Engine';
import React from 'react';
import MoleButton from '../shared/MoleButton';
import { supabase } from '@/lib/supabaseClient';
import { useRouter } from 'next/router';
import MoleModal from '../shared/MoleModal';

interface MenuModalProps {
  counter: number;
  isRunning: boolean;
  setIsRunning: (isRunning: boolean) => void;
  resetGame: () => void;
  score: number;
}

const MenuModal = ({ counter, isRunning, setIsRunning, resetGame, score }: MenuModalProps): JSX.Element => {
  const router = useRouter();

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const form = e.target as HTMLFormElement;
    const name = (form[0] as HTMLInputElement)?.value;
    const { error } = await supabase.from('Leaderboard').insert([{ name, score }]);
    if (error) {
      console.log(error);
    }
    router.push('/leaderboard');
  };

  return (
    <MoleModal>
      <h1 className="text-4xl">Whack a Mole</h1>
      {counter === 0 ? (
        <>
          <p className="text-2xl">Your score is {score}</p>
          <form onSubmit={handleSubmit} className="flex flex-col gap-4" aria-label="form">
            <input type="text" placeholder="Enter your name" className="bg-stone-600 p-2 rounded-xl" />
            <input type="submit" value="Submit" className="bg-stone-600 p-2 rounded-xl cursor-pointer" />
          </form>
        </>
      ) : (
        <>
          <MoleButton aria-label="Continue/Start" onClick={() => setIsRunning(!isRunning)}>
            {counter !== INITIAL_TIMER ? 'Continue' : 'Start'}
          </MoleButton>
          {counter !== INITIAL_TIMER && (
            <MoleButton onClick={resetGame} className="bg-stone-600 p-2 rounded-xl" aria-label="NewGame">
              Start new game
            </MoleButton>
          )}
          <MoleButton onClick={() => router.push('/leaderboard')}>Check Leaderboards</MoleButton>
        </>
      )}
    </MoleModal>
  );
};

export default MenuModal;
