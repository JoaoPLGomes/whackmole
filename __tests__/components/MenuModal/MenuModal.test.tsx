import MenuModal from '@/components/MenuModal/MenuModal';
import { render, screen, waitFor } from '@testing-library/react';

const push = jest.fn();

jest.spyOn(require('next/router'), 'useRouter').mockImplementation(() => ({
  route: '/',
  pathname: '',
  query: '',
  asPath: '',
  push,
}));

describe('MenuModal', () => {
  it('renders correct text', () => {
    render(<MenuModal counter={0} isRunning setIsRunning={jest.fn()} resetGame={jest.fn()} score={0} />);
    const text = screen.getByText('Whack a Mole');
    expect(text).toBeInTheDocument();
  });
  it('renders correct text when counter is 1', () => {
    render(<MenuModal counter={1} isRunning setIsRunning={jest.fn()} resetGame={jest.fn()} score={0} />);
    const text = screen.getByText('Continue');
    expect(text).toBeInTheDocument();
  });
  it('renders correct text when counter is 0', () => {
    render(<MenuModal counter={0} isRunning setIsRunning={jest.fn()} resetGame={jest.fn()} score={0} />);
    const text = screen.getByText('Your score is 0');
    expect(text).toBeInTheDocument();
  });

  it('calls setIsRunning when button is clicked', () => {
    const setIsRunning = jest.fn();
    render(<MenuModal counter={1} isRunning setIsRunning={setIsRunning} resetGame={jest.fn()} score={0} />);
    const button = screen.getByRole('button', { name: 'Continue/Start' });
    button.click();
    expect(setIsRunning).toHaveBeenCalled();
  });
  it('calls resetGame when button is clicked', () => {
    const resetGame = jest.fn();
    render(<MenuModal counter={1} isRunning setIsRunning={jest.fn()} resetGame={resetGame} score={0} />);
    const button = screen.getByRole('button', { name: 'NewGame' });
    button.click();
    expect(resetGame).toHaveBeenCalled();
  });
  it('submits form and redirects to leaderboard page', async () => {
    render(<MenuModal counter={0} isRunning setIsRunning={jest.fn()} resetGame={jest.fn()} score={0} />);
    const form = screen.getByRole('form', { name: 'form' }) as HTMLFormElement;
    form.submit();
    await waitFor(() => expect(push).toHaveBeenCalledWith('/leaderboard'));
  });
});
