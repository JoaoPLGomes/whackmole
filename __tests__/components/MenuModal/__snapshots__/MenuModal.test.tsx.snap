// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuModal renders correctly 1`] = `
<div>
  <div
    class="absolute bg-stone-800 text-white w-fit h-fit p-8 z-10 rounded-xl text-center flex flex-col gap-4"
  >
    <h1
      class="text-4xl"
    >
      Whack a Mole
    </h1>
    <p
      class="text-2xl"
    >
      Your score is 
      0
    </p>
    <form
      class="flex flex-col gap-4"
    >
      <input
        class="bg-stone-600 p-2 rounded-xl"
        placeholder="Enter your name"
        type="text"
      />
      <input
        class="bg-stone-600 p-2 rounded-xl cursor-pointer"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</div>
`;
